### ISSUE TITLE
CI: pass the number of retries to mocha

### ISSUE THREAD

**Issue by fflorent** [gristlabs/grist-core#938]
The CI may fail for random reasons (mostly due to timeouts). 

It becomes hard to know whether the failure is due to our additions or to the instabilities.

Mocha allows to globally set a number of retries for tests: https://mochajs.org/api/mocha#retries

May be worth for nbrowser tests (AFAIK, these are the most unstable). WDYT?

**Comment by paulfitz** [comment]
I've been wanting to look at what the tests are doing, during the obvious and common timeouts. Is it somehow downloading some big silly webdriver binary, or what is going on? Ideally someone would take some time to understand what exactly is timing out before wrapping it in retries.

**Comment by chrisdebian** [comment]
Picking this back up since it's been quiet a couple of years, with some data that might help answer @paulfitz's question about what's actually timing out.

I looked at 9 recent failing runs of the `CI` workflow on `main` (2026-08-14 through 2026-08-28). Failures aren't concentrated in one nbrowser shard or one test file, they're spread across nearly every matrix group:

- `nbrowser-^[AGIO]`: 5 runs (AdminPanel, Importer, AccessRules2 in different runs)
- `nbrowser-^[RSTU]`: 3 runs
- the catch-all `nbrowser-^[^ACFGIMORSTU]` group: 3 runs
- `nbrowser-^[FM]`, `nbrowser-^C`, `server-1-of-2`: 1 run each

Each failing run tends to hit different, unrelated test suites (AdminPanel, AccessRules2, Importer, others), which reads more like resource contention or timing pressure during the actual browser-driven test run than a specific bad test.

One concrete data point: PR #2244 tried to fix `AccessRules2` flakiness in April, but the description itself notes it couldn't be reproduced locally after 20 runs, so there was no way to confirm the fix actually worked. That same test (`should support user-attribute rules`) failed again in one of the runs I looked at on 2026-08-28, with a plausible related cause, a stale `.getAclResources()` response landing after the assertion ran.

I also checked whether this could be a Chrome/chromedriver version mismatch, since that's a common cause of exactly this kind of scattered flakiness. It doesn't look like it: `buildtools/install_chrome_for_tests.sh` pins Chrome and resolves chromedriver via `selenium-manager`, which should keep them in sync. Worth noting as a secondary thing though: every one of the ~13 matrix jobs downgrades the runner's preinstalled Chrome via its own ~127MB download, all in parallel, on every run. That finishes well before the actual test failures start, so I don't think it's the direct cause, but it's real shared overhead on a runner class that's already running many concurrent browser instances.

Not trying to propose a fix here, just picking up the "understand what's timing out" thread since nobody had. Happy to dig into a specific shard or test further if that's useful, but figured the pattern across runs was worth surfacing before guessing at a fix.

**Comment by paulfitz** [comment]
Hi @chrisdebian, thanks for looking at it. It'd be need to get that chrome load cached. We've whack-a-moled a bunch of test problems since, although more remain. One approach I've found helpful recently is to remember to add more signals in the app that tests can wait on, to reliably determine conditions that are otherwise race-y (e.g. https://github.com/gristlabs/grist-core/pull/2498). Touching the app while writing individual tests can feel "out of scope" but if no-one adds the helper conditions, tests stay bad.
